Разработка визуального компонента на базе шаблонизатора

v-nvtsk opened the pull request:
[Preview on gh-pages](https://v-nvtsk.github.io/otus-jsbasic-dz32-templater/)

[![Lint and Test](https://github.com/v-nvtsk/otus-jsbasic-dz32-templater/actions/workflows/lint-test.yaml/badge.svg?branch=develop)](https://github.com/v-nvtsk/otus-jsbasic-dz32-templater/actions/workflows/lint-test.yaml) ![Endpoint Badge](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/v-nvtsk/f9b687636482339cabd6a8c4b369f3eb/raw/1df26bb177cc8eb404b4f509e50c3cbc583db4ed/otus-jsbasic-dz32-templater-junit-tests.json) 

## Раразработать шаблонизатор
Шаблонизатор поддерживает:
- [x] подстановку данных
- [x] условный рендеринг
- [x] работу со списками
- [x] поддерживать дополнительные переменные в циклах (index, isFirst, isLast etc.)

Компонент должен поддерживать

- [x] передачу частичных свойств в конструктор
- [x] подписку на события, объявленные в свойстве events
- [x] рендеринг шаблона, который возвращается из метода render() c подстановкой данных из state
- [x] вызов хука onMount при первом рендеринге
- [x] обновление представления при изменении state через setState

![image](https://github.com/v-nvtsk/otus-jsbasic-dz32-templator/assets/138329268/8ef36a80-4711-4a61-ba68-d8d899cce3fa)

v-nvtsk committed:
build: add initial config

v-nvtsk committed:
feat: add sources translated to TS

v-nvtsk committed:
feat: add templater

v-nvtsk committed:
feat: add base-component

v-nvtsk committed:
feat: add weather component

v-nvtsk committed:
feat: fix entry point index.ts, add initial tests

v-nvtsk committed:
Update package.json

update app name

vvscode commented on src/components/base-component/base-component.ts:
```diff
@@ -0,0 +1,55 @@
+    this.onMount(el);
+  }
+
+  eventListener() {
```
методы - действия - называем глаголами

vvscode replied:
и я не понял, где он вызывается

я НЕ хочу подписываться на события руками, я хочу просто создать свойство в классе-наследнике и дальше пускай работает

v-nvtsk replied:
Исправлю

vvscode commented on src/components/base-component/base-component.ts:
```diff
@@ -0,0 +1,55 @@
+      this.el.addEventListener(action, (ev: Event) => {
+        if (ev.target !== null && ev.target instanceof HTMLElement) {
+          let currentSelector = ev.target!.tagName;
+          const currentId = ev.target!.id === "" ? "" : `#${ev.target!.id}`;
```
круто, что попробовал это сделать
но
https://developer.mozilla.org/ru/docs/Web/API/Element/matches

v-nvtsk replied:
Спасибо, искал что-то такое, но не нашёл. Попробую исправить

vvscode commented on src/components/weather/weather.ts:
```diff
@@ -0,0 +1,104 @@
+
+    const { currentCity, currentTemp, currentIcon, coord } = this.state;
+    const cities = [...citiesCache.cities];
+    return template(templateStr, { currentCity, currentTemp, currentIcon, cities, coord });
```
это должно быть внутри BaseComponent

v-nvtsk replied:
Не понимаю как это будет работать

v-nvtsk replied:
Сделал

vvscode commented on src/components/weather/weather.ts:
```diff
@@ -0,0 +1,104 @@
+  constructor(el: HTMLElement, initialState?: Partial<State>) {
+    super(el, initialState);
+    citiesCache.init();
+    citiesCache.subscribe(() => {
```
и тут у тебя, кажется, потекла память

v-nvtsk replied:
Почему? Вопрос в отсутствии возможности отписаться?

vvscode replied:
да

v-nvtsk replied:
Добавил отписку. И метод onUnmount в котором отписываюсь.

vvscode commented on src/components/base-component/base-component.ts:
```diff
@@ -0,0 +1,55 @@
+  protected onMount(el: HTMLElement) {
+    // useless activity
+    const { innerHTML } = el;
+    this.el.innerHTML = innerHTML;
```
вот это я вообще не понял, для чего =)

v-nvtsk replied:
Заглушка, потому что я так и не понял зачем это нужно.
Typescript требовал обратиться к чему-нибудь через this

vvscode replied:
ну ты этим просто взял и переписал все ноды - давай лучше пустым метод оставим (ну или прочитаем значение, но писать его не будем)

vvscode replied:
и зачем аргумент, если есть `this.el` ?

v-nvtsk replied:
> ну ты этим просто взял и переписал все ноды - давай лучше пустым метод оставим (ну или прочитаем значение, но писать его не будем)

Согласен. Убрал.

> и зачем аргумент, если есть `this.el` ?

Так вот и я не понимал что надо здесь сделать - зачем el, мне ведь всё-равно надо хранить this.el.
Или его не надо хранить?

Перенес в метод onMount вызов функции добавления слушателей, раз уж мы его каждый рендер вызываем.

vvscode commented on src/components/weather/weather.test.ts:
```diff
@@ -0,0 +1,142 @@
+  });
+
+  beforeEach(async () => {
+    global.fetch = jest.fn().mockImplementation((url: string) => {
```
а зачем так сложно? почему просто строку не закинуть?

v-nvtsk replied:
я мокаю fetch, который располагается в модулях. Для них и имитирую нормальную работу fetch.
Наверное, можно было упростить - не помню почему так всё усложнилось, но что-то отказывалось работать.

vvscode replied:
про jest я понимаю

просто тут ты делаешь 4 разных варианта мока, и это
1) сложно читать
2) сложно помнить

я советую в каждом тесте (раз ответы у тебя разные)
делать моки (а чтобы это было проще - сделать вспомогательную функцию)

v-nvtsk replied:
Разнёс моки по тестам.

vvscode commented on src/components/weather/weather.ts:
```diff
@@ -0,0 +1,104 @@
+          this.eventListener();
+        });
+    } else {
+      this.updateWeather(this.state.currentCity).then(() => {
```
кажется, тут одинаковая логика в ветках - давай ее объеденим, сможешь?

v-nvtsk replied:
сделал

v-nvtsk committed:
refactor: remove duplicating code blocks

v-nvtsk committed:
refactor: weather

trim cityName on submit
add button type="submit"
change submit event from click to form.submit
change uppercase tags in event@selectors
fix if-endif for currentTemp in template
change currentTemp type to string

v-nvtsk committed:
refactor: change getWeather return temp type to string

v-nvtsk committed:
fix: handle conditions and patterns templater

handle zero vaues for patternMatcher substitution
clear code from comments

v-nvtsk committed:
build: remove conflicting rule indent

v-nvtsk committed:
refactor: remove geoIp call from entry point

v-nvtsk committed:
refactor: base-component

refactor function eventHandlers
rename this function to setEventHandlers

v-nvtsk committed:
refactor: templater

add ability to use nested objects as conditional
fix links to regex101 samples

v-nvtsk committed:
refactor: base-component

setstate and init eventhandlers as async
remove excessive if's
fix re-render in setState

v-nvtsk committed:
refactor: weather

remove repeating render on citiesList update
render now just returns template (any) string
templater moved to BaseComponent

vvscode commented on src/components/base-component/base-component.ts:
```diff
@@ -0,0 +1,47 @@
+    Object.entries(this.events).forEach(([key, callback]) => {
+      const [action, selector] = key.split("@");
+      this.el.addEventListener(action, (ev: Event) => {
+        if ((ev.target as HTMLElement).matches(selector)) callback(ev);
```
Просто имей в виду, что не все события всплывают https://stackoverflow.com/questions/5574207/html-dom-which-events-do-not-bubble

v-nvtsk replied:
Хорошо. Буду иметь ввиду.
Переделал.

v-nvtsk committed:
refactor: BaseComponent

re-render only if state changed
listeners are added directly to elements
fix: prevent onMount method from rewrtite all nodes
default render method returns empty string

v-nvtsk committed:
refactor: CitiesCache

add function createCitiesCache returning object CitiesCache
add unsubscribe method
fix return type for addCity

v-nvtsk committed:
refactor: Weather component

add onUnmout method to unsubscribe from citiesList
add listener for citiesList instead of arrow function
 to be possible to unsubsribe from it

vvscode commented on src/components/weather/weather.test.ts:
```diff
@@ -0,0 +1,155 @@
+    await sleep();
+    expect(el.querySelectorAll(".savedCity").length).toBe(2);
+
+    cityInput = el.querySelector("input#inputCity") as HTMLInputElement;
```
ну и хорошо бы на уровне тестов оперировать не dom-элементами, а действиями понятными человеку

v-nvtsk replied:
создать вспомогательные функции?

v-nvtsk replied:
Сделал

v-nvtsk committed:
refactor: CitiesCache is class now

v-nvtsk committed:
refactor: storage module moved to citiesCache dir

v-nvtsk committed:
refactor: weather

minor changes to Weather:
 changes because of CitiesCache is now class
refactor weather tests
 add utility functions
 removed DOM calls from tests
 mock functions moved to every test block

v-nvtsk committed:
style: clear from comments

v-nvtsk committed:
refactor: citiesCache

fix: prevent setCities add list exceeding 10 items
add tests for initial loadItems
add tests for subscribe and unsubscribe
add trigger method

v-nvtsk committed:
refactor: Weather component

remove needless unMount call in method init,
  it's called from BaseComponent constructor
move mock implementations to functions

vvscode commented on .github/workflows/deploy-gh-pages.yaml:
```diff
@@ -0,0 +1,37 @@
+      # - name: Test
+      #   run: npm test
+
+      # - name: CI Badges
```
давай чистить код, а не комментировать

v-nvtsk replied:
Ок.

vvscode commented on src/CitiesCache/cities-cache.test.ts:
```diff
@@ -0,0 +1,116 @@
+
+  beforeEach(() => {
+    localStorage.clear();
+    jest.restoreAllMocks();
```
с https://github.com/v-nvtsk/otus-jsbasic-dz32-templater/pull/1/files#diff-860bd1f15d1e0bafcfc6f62560524f588e6d6bf56d4ab1b0f6f8146461558730R21 это не обязательно (и лучше это в afterEach )

v-nvtsk replied:
У меня была проблема, что после первого теста, где я мокаю localStorage - он не восстанавливается. Поэтому сначала добавил restoreAllMocks в конце теста (ваше следующее замечание), а потом перенес сюда, но забыл удалить в первом тесте.
На счёт afterEach замечание учёл, перенёс в него.

v-nvtsk replied:
Ещё один момент - clearMocks не заменяет restoreMocks

vvscode commented on src/CitiesCache/cities-cache.test.ts:
```diff
@@ -0,0 +1,116 @@
+    const testList = ["1", "2", "3", "4"];
+    mockedGetItem = jest.spyOn(storage, "getLocalStorageItem").mockReturnValue(testList);
+    expect(citiesCache.loadItems()).toEqual(testList);
+    jest.restoreAllMocks();
```
зачем это в тесте делать?

v-nvtsk replied:
Спасибо. Убрал.
Добавлял, когда тесты непонятно почему ломались.
На самом деле он только здесь и нужен только один раз.

vvscode commented on src/CitiesCache/cities-cache.ts:
```diff
@@ -0,0 +1,55 @@
+    this.cities = [...cities].slice(0, 10);
+  }
+
+  getCities() {
```
ты же уже делал задание по разработке апи хранилища - давай перейдем на асинхронный интерфейс (а то и вообще можно было бы заиспользовать реализацию из другой домашки)

vvscode commented on src/api/get-weather.ts:
```diff
@@ -0,0 +1,25 @@
+  let temp;
+  let icon;
+  let coord: [number, number];
+  if (data.cod === 200) {
```
https://developer.mozilla.org/en-US/docs/Web/API/Response/ok

v-nvtsk replied:
Не обратил внимание, что API возвращает 404, если не находит город. Думал, что всегда возвращает 200, а в поле cod уже статус.
Исправил.

vvscode commented on src/api/get-weather.ts:
```diff
@@ -0,0 +1,25 @@
+    `https://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${API_KEY}&units=${units}&lang=${lang}`,
+  );
+  const data = await response.json();
+  let city;
```
я не понял зачем эти переменные отдельно и почему не сделать `return` внутри ветки (тогда и else не понадобится) - вкусовщина, но все же вопросы вызывает

v-nvtsk replied:
После устранения предыдущего замечания - больше не нужны. Убрал.

vvscode commented on src/api/ip.test.ts:
```diff
@@ -0,0 +1,46 @@
+      }),
+    );
+
+    result = await getCityByIP();
```
вот эту идею я не понял - зачем это в beforeEach?

v-nvtsk replied:
Убрал.
Тоже не понял)

vvscode commented on src/api/ip.test.ts:
```diff
@@ -0,0 +1,46 @@
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
```
а для чего это тут? оно не поможет с ручным моком (как на 25й строке) + у тебя в конфиге это уже включено

v-nvtsk replied:
Тоже не понял. Убрал

vvscode commented on src/api/ip.test.ts:
```diff
@@ -0,0 +1,46 @@
+  });
+
+  it("getCityByIP is a function", () => {
+    expect(getCityByIP).toBeInstanceOf(Function);
```
эти тесты мы писали только в начале, для проверки окружения, в реальности он уже включен в тест ниже

v-nvtsk replied:
То есть такие тесты вообще не писать?

v-nvtsk committed:
fix: config gh-pages action

uncomment badges flow
remove commented tests flow, it presented in other action

v-nvtsk committed:
fix: clear citiesCache from needless calls restoreMock

v-nvtsk committed:
refactor: get-weather

remove API returned json.cod check
add response.ok check
removed needless variables
add field ok:true to mocks in tests

v-nvtsk committed:
refactor get-weather

remove needless mock in test
remove cod var in return-type

v-nvtsk committed:
refactor: clear ip-test from needless code
